plutus-core/untyped-plutus-core/src/UntypedPlutusCore/NewParser.hs

```diff
@@ -0,0 +1,146 @@
+import           Text.Megaparsec                 hiding (ParseError, State, parse)
+import           UntypedPlutusCore.Check.Uniques (checkProgram)
+import           UntypedPlutusCore.Rename        (Rename (rename))
+
+import           Data.ByteString.Lazy            (ByteString)
+import           Data.ByteString.Lazy.Internal   (unpackChars)
+import qualified Data.Text                       as T
+import           PlutusCore.ParserCommon
+
+-- A small type wrapper for parsers that are parametric in the type of term they parse
+type Parametric name uni fun
+    = forall term. UPLC.TermLike term PLC.TyName name uni fun
+    => Parser (term SourcePos)
+    -> Parser (term SourcePos)
+
+-- The following functions correspond to UntypedPlutusCore.Core.Type TermLike instances
+
+absTerm :: ParsecT ParseError
+                   T.Text
+                   (StateT ParserState PLC.Quote)
+                   (UPLC.Term PLC.Name uni fun SourcePos)
+                   -> Parser (UPLC.Term PLC.Name uni fun SourcePos)
+absTerm tm = UPLC.Delay <$> reservedWord "abs" <*> tm
+
+lamTerm :: ParsecT ParseError
+                   T.Text
+                   (StateT ParserState PLC.Quote)
+                   (UPLC.Term PLC.Name uni fun SourcePos)
+                   -> Parser (UPLC.Term PLC.Name uni fun SourcePos)
+lamTerm tm = UPLC.LamAbs <$> reservedWord "lam" <*> name <*> tm
+
+appTerm :: ParsecT ParseError
+                   T.Text
+                   (StateT ParserState PLC.Quote)
+                   (UPLC.Term PLC.Name uni fun SourcePos)
+                   -> Parser (UPLC.Term PLC.Name uni fun SourcePos)
+                --    PLC.Parsable (PLC.Some uni) => Parametric name uni fun
+appTerm tm = UPLC.Apply <$> getSourcePos <*> tm <*> tm
+
+conTerm
+    :: (PLC.Parsable (PLC.Some uni), PLC.Closed uni, uni `PLC.Everywhere` PLC.Parsable)
+    => Parser (UPLC.Term PLC.Name uni fun SourcePos)
+conTerm = UPLC.Constant <$> reservedWord "con" <*> constant
+
+builtinTerm :: (Bounded fun, Enum fun, Pretty fun) => Parametric name uni fun
+builtinTerm _term = UPLC.builtin <$> reservedWord "builtin" <*> builtinFunction
+
+tyInstTerm :: ParsecT ParseError
+                   T.Text
+                   (StateT ParserState PLC.Quote)
+                   (UPLC.Term PLC.Name uni fun SourcePos)
+                   -> Parser (UPLC.Term PLC.Name uni fun SourcePos)
+tyInstTerm tm = UPLC.Force <$> getSourcePos <*> tm
+
+-- In uplc, Iwrap and Unwrap are removed.
+
+errorTerm
+    :: Parser (UPLC.Term PLC.Name uni fun SourcePos)
+errorTerm = UPLC.Error <$> reservedWord "error"
+
+term
+    :: ( PLC.Parsable (PLC.Some uni), PLC.Closed uni, uni `PLC.Everywhere` PLC.Parsable
+       , Bounded fun, Enum fun, Pretty fun
+       )
+        => Parser (UPLC.Term PLC.Name uni fun SourcePos)
+term = (var >>= (\n -> getSourcePos >>= \p -> return $ UPLC.Var p n))
+    <|> (inParens $ absTerm self <|> lamTerm self <|> conTerm <|> builtinTerm self <|> errorTerm)
```

Comment:
Hmm, I would probably have considered the parens/braces to be part of the term, and so put them inside the individual parsers. I guess this repeats less... but it would be nice if `lamTerm` actually parsed a lambda term rather than the inside of a lambda term.